src/hotspot/share/opto/printinlining.cpp

```diff
@@ -0,0 +1,105 @@
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "printinlining.hpp"
+#include "opto/callnode.hpp"
+#include "memory/allocation.hpp"
+#include "memory/resourceArea.hpp"
+
+InlinePrinter::InlinePrinter(Arena* arena, bool enabled) : _enabled(enabled), _root(new(arena) IPInlineSite(nullptr, arena, 0)) {}
+
+InlinePrinter::IPInlineAttempt::IPInlineAttempt(InliningResult result) : result(result) {}
+
+outputStream* InlinePrinter::record(ciMethod* callee, JVMState* state, InliningResult result, const char* msg) {
+  if (!_enabled) {
+    return &_nullStream;
+  }
+  IPInlineAttempt* attempt = locate(state, callee)->add(result);
+  if (msg != nullptr) {
+    attempt->msg.print("%s", msg);
+  }
+  return &attempt->msg; // IPInlineAttempts are heap allocated so this address is safe
+}
+
+void InlinePrinter::print_on(outputStream* tty) {
+  if (!_enabled) {
+    return;
+  }
+  _root->dump(tty, -1);
+}
+
+InlinePrinter::IPInlineSite* InlinePrinter::locate(JVMState* state, ciMethod* callee) {
+  if (state == nullptr) {
+    return _root;
+  }
+
+  return locate(state->caller(), nullptr)->at_bci(state->bci(), callee);
+}
+
+InlinePrinter::IPInlineSite* InlinePrinter::IPInlineSite::at_bci(int bci, ciMethod* callee) {
+  int index = (bci == -1) ? 0 : bci; // -1 is a special case for some intrinsics (e.g. java.lang.ref.reference.get)
+  assert(index >= 0, "index cannot be negative");
+  if (_children.length() <= index) {
+    assert(callee != nullptr, "an inline call is missing in the chain up to the root");
+    auto child = new (_arena) IPInlineSite(callee, _arena, bci);
+    _children.at_put_grow(index, child, nullptr);
+    return child;
+  }
+  if (IPInlineSite* child = _children.at(index)) {
+    if (callee != nullptr && callee != child->_method) {
+      IPInlineAttempt* attempt = child->add(InliningResult::SUCCESS);
+      attempt->msg.print("callee changed to ");
+      CompileTask::print_inline_inner_method_info(&attempt->msg, callee);
+    }
+    return child;
+  }
+  auto child = new (_arena) IPInlineSite(callee, _arena, bci);
```

Comment:
This code is nice and compact, but I'm worried about the memory footprint.  In the worst case, we get an array element for every bytecode parsed, right?  It might be better to use a hash table.